src/Aspire.Hosting.Azure/CdkResource.cs

```diff
@@ -0,0 +1,112 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting.Lifecycle;
+using Azure.Provisioning;
+using Azure.Provisioning.ResourceManager;
+using AspireResource = Aspire.Hosting.ApplicationModel.Resource;
+
+namespace Aspire.Hosting;
+
+/// <summary>
+/// An Aspire resource which is also a CDK construct.
+/// </summary>
+/// <param name="name"></param>
+public class CdkResource(string name) : AspireResource(name)
+{
+
+}
+
+/// <summary>
+/// TODO:
+/// </summary>
+/// <param name="name"></param>
+public class AzureStorageCdkResource(string name) : CdkResource(name)
+{
+
+}
+
+/// <summary>
+/// Extensions for working with CDK resources in the .NET Aspire application model.
+/// </summary>
+public static class CdkResourceExtensions
+{
+    /// <summary>
+    /// Adds a CDK resource to the application model.
+    /// </summary>
+    /// <param name="builder">The distributed application builder.</param>
+    /// <param name="name">The name of the resource being added.</param>
+    /// <param name="callback">A callback used to configure the resource.</param>
+    /// <returns></returns>
+    public static IResourceBuilder<CdkResource> AddCdkResource(this IDistributedApplicationBuilder builder, string name, Action<AspireResourceConstruct>? callback = null)
+    {
+        builder.Services.TryAddLifecycleHook<CdkLifecycleHook>();
+
+        callback ??= static (construct) => { };
+
+        var resource = new CdkResource(name);
+        return builder.AddResource(resource)
+                      .WithAnnotation(new AspireResourceConstructCallbackAnnotation(callback), ResourceAnnotationMutationBehavior.Replace);
+    }
+}
+
+/// <summary>
+/// An Azure.Provisioning construct which maps to an Aspire Resource.
+/// </summary>
+public class AspireResourceConstruct(IConstruct? scope, string name, ConstructScope constructScope = ConstructScope.ResourceGroup, Guid? tenantId = null, Guid? subscriptionId = null, string? envName = null, ResourceGroup? resourceGroup = null) : Construct(scope, name, constructScope, tenantId, subscriptionId, envName, resourceGroup)
+{
+}
+
+internal class AspireInfrastructureConstruct(string envName) : Infrastructure(envName: envName)
+{
+}
+
+/// <summary>
+/// TODO: Doc comments.
+/// </summary>
+public class AspireResourceConstructCallbackAnnotation(Action<AspireResourceConstruct> callback) : IResourceAnnotation
+{
+    /// <summary>
+    /// TODO: Doc comments.
+    /// </summary>
+    public Action<AspireResourceConstruct> Callback = callback;
+}
+
+internal class CdkLifecycleHook(DistributedApplicationExecutionContext executionContext) : IDistributedApplicationLifecycleHook
```

Comment:
Are you trying to generate a single bicep file? I would start by doing what we do today with the AzureBicepResource. Generate bicep snippets from each CdkResource separately. Work on expressing the ability to create a construct and reference the outputs in the app model.

Reply:
We don't have the defaulting mechanism for sub, tenant and RG yet so I was just using this to make progress. I think the right idea is the geneate individual Bicep files but we'll need to keep full generation in mind.